caffe2/core/dispatch/DeviceId.h

```diff
@@ -0,0 +1,28 @@
+#pragma once
+
+#include <functional>
+#include <iostream>
+
+namespace c10 {
+
+enum class DeviceId : uint8_t {
+    UNDEFINED,
```

Comment:
As pointed out on https://github.com/pytorch/pytorch/pull/8389#discussion_r194973077, existing implementations and serializations use 0 as CPU to start the DeviceId. Can we make things consistent and basically do CPU=0 here as well? I understand that having an UNDEFINED is good to enforce people to write the DeviceId, but it seems to be a big potential backward-breaking design change.

Reply:
You're not supposed to ever do `static_cast<int>(DeviceId)`. Semantically, these enum members don't have an int value assigned to them.

If serialization logic needs to cast these from/to int, then that's the responsibility of the serialization logic and they define functions to map this enum from/to int (or, maybe a better choice, from/to string)

Reply:
Or wait, are you worried about the int values or about the default behavior? Default behavior is also independent of this class and the job of the tensor constructor or factory.

Right now in the tensor prototype, this `UNDEFINED` is used to be able to create an undefined tensor when you just do

    Tensor myTensor;

This is, however, something I don't really like to be honest. I'd rather force the tensor default constructor to either not exist, or to do something valid. Not having an undefined state sounds much better to me. cc @ezyang who had some points on why we unfortunately need undefined state.

Reply:
Ah, I mean the actual underlying numbers are still going to be valuable in understanding the values. I agree that the usage of raw values is not something that one should use enigma, but mentally, non matching enum values still add a lot of overhead. Imagine one person looking at the serialized model and see “ah this device is device value 1” and then she will get confused what 1 means - in today’s pytorch and Caffe2 impl it is gpu, and in this case it is cuda.

Reply:
What I mean is that this person shouldn't look at the enum but at the serialization code to see what the number means. Adding numbers to the enum here suggests a relationship that isn't there. Nothing checks that you aren't doing anything stupid like

    enum class DeviceId : uint8_t {
        CPU = 0,
        CUDA = 1
    };
    int serialize(DeviceId id) {
       switch(id) {
         case CPU: return 1;
         case CUDA: return 0;
       }
    }

Given that, the number has exactly the same safety guarantees as a comment

    enum class DeviceId : uint8_t {
        CPU,  // this is serialized as 0
        CUDA  // this is serialized as 1
     };

And we probably wouldn't write such a comment.

Reply:
I agree with YQ - let's just make this enum match DeviceType **exactly**. We can make UNDEFINED to be -1 if necessary?

Yes, I understand the type safety guarantees, but you'd have much easier life just matching them and ease cognitive load for people reading the code.

Also, we'd need to unify this DeviceId with tensor constructors in ATen at some point (at::kCPU).